File: unity/NeshamaSDK/Runtime/SoulEngine/Behavior/BehaviorMapper.cs
using System;
using System.Collections.Generic;
using Neshama.SoulEngine.Personality;

namespace Neshama.SoulEngine.Behavior
{
    /// <summary>
    /// Maps emotion states to game behaviors.
    /// Ported from Python npc_behavior.py NPCBehaviorBridge.
    /// 
    /// Converts emotion intensities into actionable behavior modifiers.
    /// Pure rule-based, no LLM calls.
    /// </summary>
    public class BehaviorMapper
    {
        /// <summary>
        /// A single behavior modification from emotion state.
        /// </summary>
        public struct BehaviorModifier
        {
            public BehaviorType behaviorType;
            public float modifierValue;
            public bool enabled;
            public int priority;
            public string description;
        }

        /// <summary>
        /// Complete behavior profile for an NPC.
        /// Ported from Python BehaviorProfile.
        /// </summary>
        public class BehaviorProfile
        {
            public List<BehaviorModifier> modifiers = new List<BehaviorModifier>();
            public DialogueStyle dialogueStyle = DialogueStyle.Neutral;
            public MovementPattern movementPattern = MovementPattern.Normal;
            public QuestModifier questModifier = QuestModifier.Available;
            public float shopPriceMultiplier = 1.0f;
            public float factionPointModifier = 0.0f;
            public bool willTalk = true;
            public bool willShareSecrets = false;
        }

        // ── Threshold Configs ────────────────────────────────────────────────────

        /// <summary>
        /// Emotion threshold configurations.
        /// Format: (emotionName, threshold, behaviorType, baseModifier)
        /// Ported from Python EMOTION_THRESHOLD_CONFIGS.
        /// </summary>
        private static readonly (string emotion, float threshold, BehaviorType behaviorType, float baseModifier)[] DefaultThresholdConfigs =
        {
            // High anger thresholds
            ("anger", 0.7f, BehaviorType.InteractionAllowed, -0.3f),
            ("anger", 0.8f, BehaviorType.InfoSharing, -1.0f),
            // High fear thresholds
            ("fear", 0.6f, BehaviorType.MovementPatternChange, 0.0f), // fleeing
            ("fear", 0.7f, BehaviorType.InteractionAllowed, -0.4f),
            ("fear", 0.8f, BehaviorType.DialogueStyleChange, -0.5f), // submissive
            // High joy thresholds
            ("joy", 0.6f, BehaviorType.QuestAvailabilityChange, 0.3f),
            ("joy", 0.7f, BehaviorType.ShopPriceChange, -0.1f), // 10% discount
            ("joy", 0.8f, BehaviorType.InfoSharing, 0.5f),
            // High trust thresholds
            ("trust", 0.6f, BehaviorType.InfoSharing, 0.4f),
            ("trust", 0.8f, BehaviorType.QuestAvailabilityChange, 0.5f),
            ("trust", 0.8f, BehaviorType.GiftReaction, 0.5f),
            // High sadness thresholds
            ("sadness", 0.5f, BehaviorType.DialogueStyleChange, -0.3f),
            ("sadness", 0.7f, BehaviorType.QuestAvailabilityChange, -0.2f),
            ("sadness", 0.8f, BehaviorType.MovementPatternChange, 0.0f), // hiding
            // High disgust thresholds
            ("disgust", 0.5f, BehaviorType.InteractionAllowed, -0.3f),
            ("disgust", 0.7f, BehaviorType.ShopPriceChange, 0.2f), // 20% markup
            ("disgust", 0.8f, BehaviorType.InfoSharing, -0.8f),
        };

        private OCEANPersonality _personality;

        public BehaviorMapper() : this(new OCEANPersonality()) { }

        public BehaviorMapper(OCEANPersonality personality)
        {
            _personality = personality ?? new OCEANPersonality();
        }

        // ── Core Method ──────────────────────────────────────────────────────────

        /// <summary>
        /// Generate behavior profile from emotion state.
        /// Ported from Python NPCBehaviorBridge.generate_behavior().
        /// </summary>
        public BehaviorProfile GenerateBehavior(Dictionary<string, float> emotionState)
        {
            var profile = new BehaviorProfile();
            var thresholds = ApplyPersonalityModifiers();

            // Check each threshold
            foreach (var config in thresholds)
            {
                float intensity = GetValue(emotionState, config.emotion);
                if (intensity >= config.threshold)
                {
                    float overflow = intensity - config.threshold;
                    float scaledModifier = config.baseModifier * (1f + overflow);

                    profile.modifiers.Add(new BehaviorModifier
                    {
                        behaviorType = config.behaviorType,
                        modifierValue = (float)Math.Round(scaledModifier, 4),
                        enabled = true,
                        priority = (int)(intensity * 10),
                        description = $"Emotion {config.emotion} ({intensity:F2}) affects {config.behaviorType}"
                    });
                }
            }

            // Derive aggregate behaviors
            DeriveDialogueStyle(profile, emotionState);
            DeriveMovementPattern(profile, emotionState);
            DeriveQuestModifier(profile, emotionState);
            DeriveShopPrice(profile, emotionState);
            DeriveFactionShift(profile, emotionState);
            DeriveInteractionFlags(profile, emotionState);

            return profile;
        }

        // ── Personality Modifiers ────────────────────────────────────────────────

        private List<(string emotion, float threshold, BehaviorType behaviorType, float baseModifier)> ApplyPersonalityModifiers()
        {
            var modified = new List<(string, float, BehaviorType, float)>();

            foreach (var config in DefaultThresholdConfigs)
            {
                float thresholdMod = config.threshold;

                // High agreeableness = lower thresholds for positive behaviors
                if (config.behaviorType == BehaviorType.InfoSharing || 
                    config.behaviorType == BehaviorType.QuestAvailabilityChange)
                {
                    float agreeableness = _personality.agreeableness;
                    if (agreeableness > 0.6f)
                        thresholdMod = thresholdMod * (1f - (agreeableness - 0.6f) * 0.3f);
                }

                // High neuroticism = lower thresholds for fear/anger responses
                if (config.behaviorType == BehaviorType.MovementPatternChange || 
                    config.behaviorType == BehaviorType.InteractionAllowed)
                {
                    float neuroticism = _personality.neuroticism;
                    if (neuroticism > 0.6f)
                        thresholdMod = thresholdMod * (1f - (neuroticism - 0.6f) * 0.2f);
                }

                modified.Add((config.emotion, thresholdMod, config.behaviorType, config.baseModifier));
            }

            return modified;
        }

        // ── Derivation Methods ───────────────────────────────────────────────────

        private void DeriveDialogueStyle(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float anger = GetValue(emotions, "anger");
            float fear = GetValue(emotions, "fear");
            float joy = GetValue(emotions, "joy");
            float sadness = GetValue(emotions, "sadness");

            if (anger > 0.5f) profile.dialogueStyle = DialogueStyle.Aggressive;
            else if (fear > 0.5f) profile.dialogueStyle = DialogueStyle.Submissive;
            else if (joy > 0.5f) profile.dialogueStyle = DialogueStyle.Excited;
            else if (sadness > 0.4f) profile.dialogueStyle = DialogueStyle.Gloomy;
            else if (anger > 0.3f || fear > 0.3f) profile.dialogueStyle = DialogueStyle.Cautious;
            else profile.dialogueStyle = DialogueStyle.Neutral;
        }

        private void DeriveMovementPattern(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float fear = GetValue(emotions, "fear");
            float anger = GetValue(emotions, "anger");
            float joy = GetValue(emotions, "joy");
            float sadness = GetValue(emotions, "sadness");

            if (fear > 0.6f) profile.movementPattern = MovementPattern.Fleeing;
            else if (sadness > 0.6f) profile.movementPattern = MovementPattern.Hiding;
            else if (anger > 0.6f) profile.movementPattern = MovementPattern.AggressivePatrol;
            else if (fear > 0.4f || anger > 0.4f) profile.movementPattern = MovementPattern.Defensive;
            else if (joy > 0.5f) profile.movementPattern = MovementPattern.Excited;
            else profile.movementPattern = MovementPattern.Normal;
        }

        private void DeriveQuestModifier(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float anger = GetValue(emotions, "anger");
            float trust = GetValue(emotions, "trust");
            float sadness = GetValue(emotions, "sadness");

            if (anger > 0.7f) profile.questModifier = QuestModifier.Locked;
            else if (trust > 0.7f) profile.questModifier = QuestModifier.Available;
            else if (sadness > 0.5f) profile.questModifier = QuestModifier.AvailableWithCondition;
            else profile.questModifier = QuestModifier.Available;
        }

        private void DeriveShopPrice(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float joy = GetValue(emotions, "joy");
            float trust = GetValue(emotions, "trust");
            float anger = GetValue(emotions, "anger");
            float disgust = GetValue(emotions, "disgust");

            float multiplier = 1.0f;

            // Positive emotions = discount
            if (joy > 0.5f) multiplier -= 0.1f * joy;
            if (trust > 0.5f) multiplier -= 0.1f * trust;

            // Negative emotions = markup
            if (anger > 0.4f) multiplier += 0.15f * anger;
            if (disgust > 0.4f) multiplier += 0.2f * disgust;

            profile.shopPriceMultiplier = Math.Max(0.5f, Math.Min(2.0f, multiplier));
        }

        private void DeriveFactionShift(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float anger = GetValue(emotions, "anger");
            float disgust = GetValue(emotions, "disgust");
            float trust = GetValue(emotions, "trust");
            float joy = GetValue(emotions, "joy");

            float shift = 0.0f;
            if (trust > 0.4f) shift += 0.1f * trust;
            if (joy > 0.4f) shift += 0.1f * joy;
            if (anger > 0.4f) shift -= 0.15f * anger;
            if (disgust > 0.4f) shift -= 0.1f * disgust;

            profile.factionPointModifier = Math.Max(-1f, Math.Min(1f, shift));
        }

        private void DeriveInteractionFlags(BehaviorProfile profile, Dictionary<string, float> emotions)
        {
            float anger = GetValue(emotions, "anger");
            float fear = GetValue(emotions, "fear");
            float disgust = GetValue(emotions, "disgust");
            float trust = GetValue(emotions, "trust");

            // Will talk
            if (anger > 0.8f || fear > 0.8f || disgust > 0.8f)
                profile.willTalk = false;
            else if (anger > 0.5f || fear > 0.6f || disgust > 0.5f)
                profile.willTalk = false;
            else
                profile.willTalk = true;

            // Will share secrets
            if (trust > 0.7f) profile.willShareSecrets = true;
            else if (anger > 0.3f || fear > 0.4f) profile.willShareSecrets = false;
            else profile.willShareSecrets = false;
        }

        private static float GetValue(Dictionary<string, float> dict, string key)
        {
            float val;
            return dict != null && dict.TryGetValue(key, out val) ? val : 0f;
        }
    }
}

File: unreal/NeshamaSDK/Source/NeshamaSDK/Private/SoulEngine/Behavior/BehaviorMapper.cpp
#include "SoulEngine/Behavior/BehaviorMapper.h"
#include "SoulEngine/Personality/OCEANPersonality.h"
#include "SoulEngine/Utils/SoulMathUtils.h"

UBehaviorMapper::UBehaviorMapper()
	: Personality(nullptr)
{
}

void UBehaviorMapper::Initialize(UOCEANPersonality* InPersonality)
{
	Personality = InPersonality;
}

const TArray<UBehaviorMapper::FThresholdConfig>& UBehaviorMapper::GetDefaultThresholdConfigs()
{
	static TArray<FThresholdConfig> Configs;
	if (Configs.Num() > 0) return Configs;

	// High anger thresholds
	Configs.Add({TEXT("anger"), 0.7f, EBehaviorType::InteractionAllowed, -0.3f});
	Configs.Add({TEXT("anger"), 0.8f, EBehaviorType::InfoSharing, -1.0f});
	// High fear thresholds
	Configs.Add({TEXT("fear"), 0.6f, EBehaviorType::MovementPatternChange, 0.0f}); // fleeing
	Configs.Add({TEXT("fear"), 0.7f, EBehaviorType::InteractionAllowed, -0.4f});
	Configs.Add({TEXT("fear"), 0.8f, EBehaviorType::DialogueStyleChange, -0.5f}); // submissive
	// High joy thresholds
	Configs.Add({TEXT("joy"), 0.6f, EBehaviorType::QuestAvailabilityChange, 0.3f});
	Configs.Add({TEXT("joy"), 0.7f, EBehaviorType::ShopPriceChange, -0.1f}); // 10% discount
	Configs.Add({TEXT("joy"), 0.8f, EBehaviorType::InfoSharing, 0.5f});
	// High trust thresholds
	Configs.Add({TEXT("trust"), 0.6f, EBehaviorType::InfoSharing, 0.4f});
	Configs.Add({TEXT("trust"), 0.8f, EBehaviorType::QuestAvailabilityChange, 0.5f});
	Configs.Add({TEXT("trust"), 0.8f, EBehaviorType::GiftReaction, 0.5f});
	// High sadness thresholds
	Configs.Add({TEXT("sadness"), 0.5f, EBehaviorType::DialogueStyleChange, -0.3f});
	Configs.Add({TEXT("sadness"), 0.7f, EBehaviorType::QuestAvailabilityChange, -0.2f});
	Configs.Add({TEXT("sadness"), 0.8f, EBehaviorType::MovementPatternChange, 0.0f}); // hiding
	// High disgust thresholds
	Configs.Add({TEXT("disgust"), 0.5f, EBehaviorType::InteractionAllowed, -0.3f});
	Configs.Add({TEXT("disgust"), 0.7f, EBehaviorType::ShopPriceChange, 0.2f}); // 20% markup
	Configs.Add({TEXT("disgust"), 0.8f, EBehaviorType::InfoSharing, -0.8f});

	return Configs;
}

TArray<UBehaviorMapper::FThresholdConfig> UBehaviorMapper::ApplyPersonalityModifiers()
{
	TArray<FThresholdConfig> Modified;

	for (const FThresholdConfig& Config : GetDefaultThresholdConfigs())
	{
		float ThresholdMod = Config.Threshold;

		// High agreeableness = lower thresholds for positive behaviors
		if (Config.BehaviorType == EBehaviorType::InfoSharing ||
			Config.BehaviorType == EBehaviorType::QuestAvailabilityChange)
		{
			if (Personality)
			{
				float Agreeableness = Personality->Agreeableness;
				if (Agreeableness > 0.6f)
					ThresholdMod = ThresholdMod * (1.0f - (Agreeableness - 0.6f) * 0.3f);
			}
		}

		// High neuroticism = lower thresholds for fear/anger responses
		if (Config.BehaviorType == EBehaviorType::MovementPatternChange ||
			Config.BehaviorType == EBehaviorType::InteractionAllowed)
		{
			if (Personality)
			{
				float Neuroticism = Personality->Neuroticism;
				if (Neuroticism > 0.6f)
					ThresholdMod = ThresholdMod * (1.0f - (Neuroticism - 0.6f) * 0.2f);
			}
		}

		Modified.Add({Config.Emotion, ThresholdMod, Config.BehaviorType, Config.BaseModifier});
	}

	return Modified;
}

FBehaviorProfile UBehaviorMapper::GenerateBehavior(const TMap<FString, float>& EmotionState)
{
	FBehaviorProfile Profile;
	TArray<FThresholdConfig> Thresholds = ApplyPersonalityModifiers();

	// Check each threshold
	for (const FThresholdConfig& Config : Thresholds)
	{
		float Intensity = GetEmoValue(EmotionState, Config.Emotion);
		if (Intensity >= Config.Threshold)
		{
			float Overflow = Intensity - Config.Threshold;
			float ScaledModifier = Config.BaseModifier * (1.0f + Overflow);

			FBehaviorModifier Mod;
			Mod.BehaviorType = Config.BehaviorType;
			Mod.ModifierValue = NeshamaSoul::SoulMathUtils::Round(ScaledModifier);
			Mod.bEnabled = true;
			Mod.Priority = static_cast<int32>(Intensity * 10);
			Mod.Description = FString::Printf(TEXT("Emotion %s (%.2f) affects %d"),
				*Config.Emotion, Intensity, static_cast<int32>(Config.BehaviorType));
			Profile.Modifiers.Add(Mod);
		}
	}

	// Derive aggregate behaviors
	DeriveDialogueStyle(Profile, EmotionState);
	DeriveMovementPattern(Profile, EmotionState);
	DeriveQuestModifier(Profile, EmotionState);
	DeriveShopPrice(Profile, EmotionState);
	DeriveFactionShift(Profile, EmotionState);
	DeriveInteractionFlags(Profile, EmotionState);

	return Profile;
}

void UBehaviorMapper::DeriveDialogueStyle(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Fear = GetEmoValue(Emotions, TEXT("fear"));
	float Joy = GetEmoValue(Emotions, TEXT("joy"));
	float Sadness = GetEmoValue(Emotions, TEXT("sadness"));

	if (Anger > 0.5f) Profile.DialogueStyle = EDialogueStyle::Aggressive;
	else if (Fear > 0.5f) Profile.DialogueStyle = EDialogueStyle::Submissive;
	else if (Joy > 0.5f) Profile.DialogueStyle = EDialogueStyle::Excited;
	else if (Sadness > 0.4f) Profile.DialogueStyle = EDialogueStyle::Gloomy;
	else if (Anger > 0.3f || Fear > 0.3f) Profile.DialogueStyle = EDialogueStyle::Cautious;
	else Profile.DialogueStyle = EDialogueStyle::Neutral;
}

void UBehaviorMapper::DeriveMovementPattern(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Fear = GetEmoValue(Emotions, TEXT("fear"));
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Joy = GetEmoValue(Emotions, TEXT("joy"));
	float Sadness = GetEmoValue(Emotions, TEXT("sadness"));

	if (Fear > 0.6f) Profile.MovementPattern = EMovementPattern::Fleeing;
	else if (Sadness > 0.6f) Profile.MovementPattern = EMovementPattern::Hiding;
	else if (Anger > 0.6f) Profile.MovementPattern = EMovementPattern::AggressivePatrol;
	else if (Fear > 0.4f || Anger > 0.4f) Profile.MovementPattern = EMovementPattern::Defensive;
	else if (Joy > 0.5f) Profile.MovementPattern = EMovementPattern::Excited;
	else Profile.MovementPattern = EMovementPattern::Normal;
}

void UBehaviorMapper::DeriveQuestModifier(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Trust = GetEmoValue(Emotions, TEXT("trust"));
	float Sadness = GetEmoValue(Emotions, TEXT("sadness"));

	if (Anger > 0.7f) Profile.QuestModifier = EQuestModifier::Locked;
	else if (Trust > 0.7f) Profile.QuestModifier = EQuestModifier::Available;
	else if (Sadness > 0.5f) Profile.QuestModifier = EQuestModifier::AvailableWithCondition;
	else Profile.QuestModifier = EQuestModifier::Available;
}

void UBehaviorMapper::DeriveShopPrice(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Joy = GetEmoValue(Emotions, TEXT("joy"));
	float Trust = GetEmoValue(Emotions, TEXT("trust"));
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Disgust = GetEmoValue(Emotions, TEXT("disgust"));

	float Multiplier = 1.0f;

	// Positive emotions = discount
	if (Joy > 0.5f) Multiplier -= 0.1f * Joy;
	if (Trust > 0.5f) Multiplier -= 0.1f * Trust;

	// Negative emotions = markup
	if (Anger > 0.4f) Multiplier += 0.15f * Anger;
	if (Disgust > 0.4f) Multiplier += 0.2f * Disgust;

	Profile.ShopPriceMultiplier = FMath::Clamp(Multiplier, 0.5f, 2.0f);
}

void UBehaviorMapper::DeriveFactionShift(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Disgust = GetEmoValue(Emotions, TEXT("disgust"));
	float Trust = GetEmoValue(Emotions, TEXT("trust"));
	float Joy = GetEmoValue(Emotions, TEXT("joy"));

	float Shift = 0.0f;
	if (Trust > 0.4f) Shift += 0.1f * Trust;
	if (Joy > 0.4f) Shift += 0.1f * Joy;
	if (Anger > 0.4f) Shift -= 0.15f * Anger;
	if (Disgust > 0.4f) Shift -= 0.1f * Disgust;

	Profile.FactionPointModifier = FMath::Clamp(Shift, -1.0f, 1.0f);
}

void UBehaviorMapper::DeriveInteractionFlags(FBehaviorProfile& Profile, const TMap<FString, float>& Emotions)
{
	float Anger = GetEmoValue(Emotions, TEXT("anger"));
	float Fear = GetEmoValue(Emotions, TEXT("fear"));
	float Disgust = GetEmoValue(Emotions, TEXT("disgust"));
	float Trust = GetEmoValue(Emotions, TEXT("trust"));

	// Will talk
	if (Anger > 0.8f || Fear > 0.8f || Disgust > 0.8f)
		Profile.bWillTalk = false;
	else if (Anger > 0.5f || Fear > 0.6f || Disgust > 0.5f)
		Profile.bWillTalk = false;
	else
		Profile.bWillTalk = true;

	// Will share secrets
	if (Trust > 0.7f) Profile.bWillShareSecrets = true;
	else if (Anger > 0.3f || Fear > 0.4f) Profile.bWillShareSecrets = false;
	else Profile.bWillShareSecrets = false;
}

float UBehaviorMapper::GetEmoValue(const TMap<FString, float>& Dict, const FString& Key)
{
	const float* Val = Dict.Find(Key);
	return Val ? *Val : 0.0f;
}
